src/hex/edge.rs

```diff
@@ -0,0 +1,280 @@
+//! Handles edges in a hexagonal grid.
+
+// use crate::vertex;
+
+use super::{
+    coordinate::{axial, Axial, HexDirection},
+    vertex::{vertex, Vertex, VertexSpin},
+};
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+/// Orientation of an edge.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(PartialEq, Eq, Copy, Clone, Hash, Debug)]
+pub enum EdgeDirection {
+    /// West of the hex.
+    West,
+    /// North West of the hex
+    NorthWest,
+    /// North East of the hex
+    NorthEast,
+}
+
+/// A hexagonal edge.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(PartialEq, Eq, Copy, Clone, Hash, Debug)]
+pub struct Edge {
+    /// q (x) coordinate
+    pub q: i32,
+    /// r (y) coordinate
+    pub r: i32,
+    /// Edge direction
+    pub dir: EdgeDirection,
+}
+
+/// Helper macro to create [`Edge`] structs.
+#[macro_export]
+macro_rules! edge {
+    ($q:expr, $r:expr, $dir:expr) => {
+        Edge {
+            q: $q,
+            r: $r,
+            dir: $dir,
+        }
+    };
+}
+pub use edge;
+
+impl Default for Edge {
+    fn default() -> Self {
+        Self {
+            q: 0,
+            r: 0,
+            dir: EdgeDirection::West,
+        }
+    }
+}
+
+impl From<HexDirection> for Edge {
+    fn from(value: HexDirection) -> Self {
```

Comment:
There's only one significant thought I have re: this implementation, and that's that I guess there'd be two ways to implement this: the one you've chosen where there's three directions and four different possible starting locations, and one where there were six directions and a single possible starting location.

Are there specific advantages to either of them?

Reply:
I believe you are referring to whether or not to have HexDir and EdgeDir be the same? I personally think even if they look the same should remain separate since they represent different things. Same with VertexDirection. Yea HexDir and VertexDirection are 6-unit long enums defining directions, but HexDir is specific in that it MUST start in the positive q axis since logic is derived from that assumption, it could very well be assuming that it points somewhere else but why? It's also normally common practice to define a 'forward' vector for any kind of spatial math.

Also, in re to Edges specifically, what do you mean starting locations? There are 3 different edges associated to a hex coordinate, West, NorthWest, NorthEast. This is to have a shape that perfectly tiles in a hexagon grid while being mathematically tied to a coordinate and covers the entire domain of the number set -∞ => ∞. That From function can be thought of like this, how do I go from a hex coordinate (implied 0,0 here) and a direction and what edge is that referring to.

Reply:
The point re: the positive q axis makes it clear to me why it's being done this way, thanks.